src/friendly-iframe-embed.js

```diff
@@ -0,0 +1,171 @@
+ * - html: The complete content of an AMP embed, which is itself an AMP
+ *   document. Can include whatever is normally allowed in an AMP document,
+ *   except for AMP `<script>` declarations. Those should be passed as an
+ *   array of `extensionIds`.
+ * - extensionsIds: An optional array of AMP extension IDs used in this embed.
+ * - fonts: An optional array of fonts used in this embed.
+ *
+ * @typedef {{
+ *   url: string,
+ *   html: string,
+ *   extensionIds: (?Array<string>|undefined),
+ *   fonts: (?Array<string>|undefined),
+ * }}
+ */
+export let FriendlyIframeSpec;
+
+
+/**
+ * @type {boolean|undefined}
+ * @visiblefortesting
+ */
+let srcdocSupported;
+
+/**
+ * @param {boolean|undefined} val
+ * @visiblefortesting
+ */
+export function setSrcdocSupportedForTesting(val) {
+  srcdocSupported = val;
+}
+
+/**
+ * Returns `true` if the Friendly Iframes are supported.
+ * @return {boolean}
+ */
+function isSrcdocSupported() {
+  if (srcdocSupported === undefined) {
+    srcdocSupported = 'srcdoc' in HTMLIFrameElement.prototype;
+  }
+  return srcdocSupported;
+}
+
+
+/**
+ * Creates the requested "friendly iframe" embed. Returns the promise that
+ * will be resolved as soon as the embed is available. The actual
+ * initialization of the embed will start as soon as the `iframe` is added
+ * to the DOM.
+ * @param {!HTMLIFrameElement} iframe
+ * @param {!Element} container
+ * @param {!FriendlyIframeSpec} spec
+ * @return {!Promise<FriendlyIframeEmbed>}
+ */
+export function installFriendlyIframeEmbed(iframe, container, spec) {
+  const win = getTopWindow(iframe.ownerDocument.defaultView);
+  const extensions = extensionsFor(win);
+
+  iframe.style.visibility = 'hidden';
+  iframe.setAttribute('referrerpolicy', 'unsafe-url');
+
+  // Pre-load extensions.
+  if (spec.extensionIds) {
+    spec.extensionIds.forEach(
+        extensionId => extensions.loadExtension(extensionId));
+  }
+
+  // Receive the signal when iframe is ready: it's document is formed.
+  iframe.onload = () => {
+    // Chrome does not reflect the iframe readystate.
+    iframe.readyState = 'complete';
+  };
+  let readyPromise;
+  if (isSrcdocSupported()) {
+    iframe.srcdoc = spec.html;
+    // TODO(dvoytenko): Look for a way to get a faster call from here.
+    // Experiments show that the iframe's "load" event is consistently 50-100ms
+    // later than the contentWindow actually available.
+    readyPromise = loadPromise(iframe);
+    container.appendChild(iframe);
+  } else {
+    iframe.src = 'about:blank';
+    container.appendChild(iframe);
+    const childDoc = iframe.contentWindow.document;
+    childDoc.open();
+    childDoc.write(spec.html);
+    childDoc.close();
+    readyPromise = Promise.resolve();
+  }
+  return readyPromise.then(() => {
+    const childWin = iframe.contentWindow;
+    const childDoc = childWin.document;
+
+    // Add <BASE> tag.
+    childDoc.head.appendChild(createElementWithAttributes(
```

Comment:
Well, we could try, but the document is not visible yet and not functional - none of the extensions have been installed. Thus this is still a good time to do it, imo. And it's a lot easier than trying to insert it into a HTML string.